circuits/src/zk_circuits/valid_settle.rs

```diff
@@ -0,0 +1,881 @@
+//! Defines the VALID SETTLE circuit, which is proven after a match, validating that
+//! both party's secret shares have been updated properly with the result of the match
+
+use curve25519_dalek::scalar::Scalar;
+use mpc_bulletproof::{
+    r1cs::{
+        LinearCombination, Prover, R1CSProof, RandomizableConstraintSystem, Variable, Verifier,
+    },
+    BulletproofGens,
+};
+use rand_core::{CryptoRng, OsRng, RngCore};
+use serde::{Deserialize, Serialize};
+
+use crate::{
+    errors::{ProverError, VerifierError},
+    types::{
+        r#match::{CommittedMatchResult, LinkableMatchResultCommitment, MatchResultVar},
+        wallet::{WalletSecretShare, WalletSecretShareCommitment, WalletSecretShareVar},
+    },
+    zk_gadgets::{
+        comparators::{EqGadget, EqVecGadget},
+        select::CondSelectVectorGadget,
+    },
+    CommitPublic, CommitVerifier, CommitWitness, SingleProverCircuit,
+};
+
+// ----------------------
+// | Circuit Definition |
+// ----------------------
+
+/// The circuit definition of `VALID SETTLE`
+pub struct ValidSettle<const MAX_BALANCES: usize, const MAX_ORDERS: usize, const MAX_FEES: usize>;
+impl<const MAX_BALANCES: usize, const MAX_ORDERS: usize, const MAX_FEES: usize>
+    ValidSettle<MAX_BALANCES, MAX_ORDERS, MAX_FEES>
+where
+    [(); MAX_BALANCES + MAX_ORDERS + MAX_FEES]: Sized,
+{
+    /// The circuit representing `VALID SETTLE`
+    pub fn circuit<CS: RandomizableConstraintSystem>(
+        statement: ValidSettleStatementVar<MAX_BALANCES, MAX_ORDERS, MAX_FEES>,
+        witness: ValidSettleWitnessVar<MAX_BALANCES, MAX_ORDERS, MAX_FEES>,
+        cs: &mut CS,
+    ) {
+        // Select the balances received by each party
+        let mut party0_party1_received = CondSelectVectorGadget::select(
+            &[
+                witness.match_res.quote_amount,
+                witness.match_res.base_amount,
+            ],
+            &[
+                witness.match_res.base_amount,
+                witness.match_res.quote_amount,
+            ],
+            witness.match_res.direction,
+            cs,
+        );
+
+        let party0_received_amount = party0_party1_received.remove(0);
+        let party1_received_amount = party0_party1_received.remove(0);
+
+        // Constrain the wallet updates to party0's shares
+        Self::validate_balance_updates(
+            statement.party0_send_balance_index,
+            party1_received_amount.clone(),
+            statement.party0_receive_balance_index,
+            party0_received_amount.clone(),
+            &witness.party0_public_shares,
+            &statement.party0_modified_shares,
+            cs,
+        );
+
+        Self::validate_order_updates(
+            statement.party0_order_index,
+            witness.match_res.base_amount,
```

Comment:
Like, one party is sending `base_amount`, receiving `quote_amount`, and the other is sending `quote_amount`, receiving `base_amount`, right? So we can't pass in `base_amount` to `validate_order_updates` for both parties

Reply:
Oh, nvm, looks like both parties trade on the same notions of `(base, quote)` and denote the order volume in terms of `base`, using the `OrderSide` enum to indicate which side they take.

But leaving this comment in the review in case my rationale here isn't accurate

Reply:
That's correct, to make it more clear; a common pair would be `(base, quote) = (ETH, USDC)`. Both parties specify the amount of _eth_ (base) they wish to buy/sell